Clipboard refactoring

mmiecz opened the pull request:
<!-- Please describe your changes on the following line: -->
This PR removes clipboard handling from the constellation. Instead, now embedder handles it.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #23440 (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [X] These changes do not require tests because it is enough to test manually in input box, if copying and pasting still works .

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/23564)
<!-- Reviewable:end -->

mmiecz committed:
Remove not needed DummyClipboardContext

mmiecz committed:
Remove clipboard handling from constellation, and remove ScriptMessage for getting/setting clipboard

mmiecz committed:
Add Embbedder messages for setting/getting clipboard contents

mmiecz committed:
Forward to embedder message for setting/getting clipboard contents from clipboard provider.
Create clipboard context in browser.rs and handle new messages.

mmiecz committed:
Code cleanup

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @emilio (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/clipboard_provider.rs, components/constellation/constellation.rs
 * @cbrewster: components/constellation/constellation.rs
 * @paulrouget: ports/glutin/browser.rs, ports/glutin/main2.rs, ports/glutin/Cargo.toml, components/constellation/constellation.rs
 * @kichjang: components/script/clipboard_provider.rs, components/script_traits/script_msg.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

jdm commented:
```
error: unused extern crate
 --> ports/glutin/main2.rs:5:1
  |
5 | extern crate clipboard;
  | ^^^^^^^^^^^^^^^^^^^^^^^ help: remove it
  |
  = note: `-D unused-extern-crates` implied by `-D warnings`
```

jdm commented on components/script/clipboard_provider.rs:
```diff
@@ -30,24 +29,3 @@ impl ClipboardProvider for ScriptToConstellationChan {
     }
 }
-
-pub struct DummyClipboardContext {
```
It turns out that we actually need this for the tests in `tests/unit/script/textinput.rs`, so let's move it to that file instead.

mmiecz replied:
I've moved `DummyClipboardContext` to `test/unit/script/textinput.rs`.

jdm commented on ports/glutin/main2.rs:
```diff
@@ -2,14 +2,14 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
+extern crate clipboard;
```
Let's remove this.

mmiecz replied:
Done, removed.

jdm reviewed (changes requested):
Looks good!

jdm commented:
If you build `./mach build --libsimpleservo`, you should find some other places where you need to handle the new messages.

mmiecz committed:
Remove unneeded clipboard import.

mmiecz committed:
Restore DummyClipboardContext for textinput.rs tests

mmiecz committed:
Clipboard handling in libsimpleservo

jdm commented on ports/libsimpleservo/api/Cargo.toml:
```diff
@@ -7,6 +7,7 @@ edition = "2018"
 publish = false
 
 [dependencies]
+clipboard = "0.5"
```
Since libsimpleservo/api is a platform-agnostic embedding API, we should not add a platform-specific dependency like clipboard to it.

jdm commented on ports/libsimpleservo/api/src/lib.rs:
```diff
@@ -516,6 +526,28 @@ impl ServoGlue {
                     self.events.push(WindowEvent::SelectBrowser(new_browser_id));
                 },
+                EmbedderMsg::GetClipboardContents(sender) => {
+                    let contents = match self.clipboard_ctx {
```
Instead of directly manipulating the clipboard for the GetClipboardContents and SetClipboardContents messages, let's add new callbacks to HostCallbacks (`get_clipboard_contents()` can return an Option<String> and `set_clipboard_contents` can accept a String) and call those instead. See how EmbedderMsg::HistoryChanged is handled earlier.

jdm reviewed (changes requested):
Sorry for the delay; last week was very busy. One last change which will allow us to provide different clipboard solutions on other platforms like android/magic leap/hololens!

mmiecz committed:
Add HostTrait methods get_clipboard_contents set_clipboard_contents, and call them instead of directly handlind clipboard

mmiecz commented:
Ok thanks for the tips @jdm :).  I've removed direct handling of clipboard from `libsimpleservo`, and added methods to `HostTrait` as mentioned.

I am not sure if I'm doing implementation right though ( should I somewhere free memory in `get_clipboard_contents()`?: 

```   rust
 fn get_clipboard_contents(&self) -> Option<String> {
        debug!("get_clipboard_contents");
        let raw_contents = (self.0.get_clipboard_contents)();
        if raw_contents.is_null() {
            return None;
        }
        let c_str = unsafe { CStr::from_ptr(raw_contents) };
        let contents_str = c_str.to_str().expect("Can't create str");
        Some(contents_str.to_owned())
    }

    fn set_clipboard_contents(&self, contents: String) {
        debug!("set_clipboard_contents");
        let contents = CString::new(contents).expect("Can't create string");
        let contents_ptr = contents.as_ptr();
        mem::forget(contents);
        (self.0.set_clipboard_contents)(contents_ptr);
    }
```
I guess this two methods should call some external ( platform specific ) function for getting and setting system clipboard. This is just for `capi`, but this should be done for `jniapi` also, right? Call java `get/setClipboardContents` from `jniapi/lib/src.rs`?

jdm commented:
For jniapi and mlservo you can leave stub implementations, unless you are familiar enough with android that you're confident about the JNI work :)

mmiecz committed:
Handle get_clipboard_contents and set_clipboard_contents in libsimpleservo capi

mmiecz committed:
Add stub for set_clipboard_contents and  get_clipboard_contents in libsimpleservo jniapi and libmlservo

jdm commented on ports/libsimpleservo/jniapi/src/lib.rs:
```diff
@@ -457,6 +457,12 @@ impl HostTrait for HostCallbacks {
     fn on_ime_state_changed(&self, _show: bool) {}
+
+    fn get_clipboard_contents(&self) -> Option<String> {
+        unimplemented!()
```
Let's make this return None instead.

jdm commented on ports/libmlservo/src/lib.rs:
```diff
@@ -354,6 +354,12 @@ impl HostTrait for HostCallbacks {
     }
+
+    fn get_clipboard_contents(&self) -> Option<String> {
+        unimplemented!()
```
Let's make this return None instead.

mmiecz committed:
Return None instead of panicking

jdm commented:
@bors-servo r+
Thanks a lot for fixing this!

bors-servo commented:
:pushpin: Commit 1bbe92f has been approved by `jdm`

<!-- @bors-servo r=jdm 1bbe92ff760ff21cf1841a96cd687c69dd5d421f -->

bors-servo commented:
:hourglass: Testing commit 1bbe92ff760ff21cf1841a96cd687c69dd5d421f with merge b22aa8bdc887cfef8c70dcec3633b3fc8ce7b766...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/12755)

jdm commented:
@bors-servo retry
* #13480

bors-servo commented:
:bomb: Failed to start rebuilding: `Unknown error`

bors-servo commented:
:hourglass: Testing commit 1bbe92ff760ff21cf1841a96cd687c69dd5d421f with merge 3186a881d3112986f436eb5aa00007eee0d3eaf0...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/Ce6tfFKBSNGWxTYvzowmzQ)

jdm commented:
A warning that needs to be fixed:
```
[0m[1m[38;5;9merror[0m[0m[1m: unused `std::result::Result` that must be used[0m
[0m   [0m[0m[1m[38;5;12m--> [0m[0mports/libsimpleservo/api/src/lib.rs:538:21[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m[1m[38;5;12m538[0m[0m [0m[0m[1m[38;5;12m| [0m[0m                    sender.send(contents.unwrap_or("".to_owned()));[0m
[0m    [0m[0m[1m[38;5;12m| [0m[0m                    [0m[0m[1m[38;5;9m^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m    [0m[0m[1m[38;5;12m= [0m[0m[1mnote[0m[0m: `-D unused-must-use` implied by `-D warnings`[0m
[0m    [0m[0m[1m[38;5;12m= [0m[0m[1mnote[0m[0m: this `Result` may be an `Err` variant, which should be handled[0m

[0m[1m[38;5;9merror[0m[0m[1m: aborting due to previous error[0m
```

mmiecz committed:
Fix warning about unused result

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 775ff3b has been approved by `jdm`

<!-- @bors-servo r=jdm 775ff3b6235922e1bec15f28aed77fc1c7e9abdb -->

bors-servo commented:
:hourglass: Testing commit 775ff3b6235922e1bec15f28aed77fc1c7e9abdb with merge 1601e75fc33937376726d729851710ad3da4bf00...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/fzc4UEFaShC1Uj-l4mIEEg)

jdm commented:
More warnings from other platforms:
```
[0m[1m[38;5;9merror[0m[0m[1m: unused variable: `contents`[0m
[0m   [0m[0m[1m[38;5;12m--> [0m[0mports/libsimpleservo/jniapi/src/lib.rs:484:38[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m[1m[38;5;12m484[0m[0m [0m[0m[1m[38;5;12m| [0m[0m    fn set_clipboard_contents(&self, contents: String) {}[0m
[0m    [0m[0m[1m[38;5;12m| [0m[0m                                     [0m[0m[1m[38;5;9m^^^^^^^^[0m[0m [0m[0m[1m[38;5;9mhelp: consider prefixing with an underscore: `_contents`[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m    [0m[0m[1m[38;5;12m= [0m[0m[1mnote[0m[0m: `-D unused-variables` implied by `-D warnings`[0m
```
```
error: unused variable: `contents`
   --> ports/libmlservo/src/lib.rs:386:38
    |
386 |     fn set_clipboard_contents(&self, contents: String) {}
    |                                      ^^^^^^^^ help: consider prefixing with an underscore: `_contents`
    |
    = note: `-D unused-variables` implied by `-D warnings`
```

mmiecz commented:
Can I build/test other platforms somehow on my Linux machine to see and fix such failures before issuing a PR? :)

jdm commented:
The jni one you could build by following https://github.com/servo/servo/wiki/Building-for-Android, but it's not necessarily worth it. You can see all of the failures from the previous build by looking at https://treeherder.allizom.org/#/jobs?repo=servo-auto .

jdm commented:
The libmlservo can only be built by people with access to the Magic Leap SDK, which is restricted to a small set of developers right now.

mmiecz committed:
Prefix unused arguments with underscore to prevent warnings

mmiecz commented:
Thanks @jdm for explanation. I've fixed the unused arguments warnings.

jdm commented:
@bors-servo r+
Thanks!

bors-servo commented:
:pushpin: Commit 7fb2bbe has been approved by `jdm`

<!-- @bors-servo r=jdm 7fb2bbed002885fb38723f5a7c75524da7fbab0b -->

bors-servo commented:
:hourglass: Testing commit 7fb2bbed002885fb38723f5a7c75524da7fbab0b with merge e382266b2295cb3dc97526bd98025f5b9221443a...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12711), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/12773), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/SE1UQ0vtSYW9CU6p1da5lw)
Approved by: jdm
Pushing e382266b2295cb3dc97526bd98025f5b9221443a to master...
